### ISSUE TITLE
[wasm] Assertion in marshal.c:399 if GetFunctionPointerForDelegate is called more than once

### ISSUE THREAD

**Issue by lambdageek** [dotnet/runtime#38573]
Run the following code on a platform that doesn't define `MONO_ARCH_HAVE_INTERP_NATIVE_TO_MANAGED`

```csharp
using System;

public delegate int MyDelegate (int i);

public class C
{
  int p;
  public C () { p = 42; }

  public int F1 (int i) => p + i;
  public int F2 (int i) => p - i;

  public static void Main()
  {
    var c = new C ();
    MyDelegate d1 = new MyDelegate (c.F1);
    MyDelegate d2 = new MyDelegate (c.F2);

    IntPtr i1 = System.Runtime.InteropServices.Marshal.GetFunctionPointerForDelegate (d1);
    IntPtr i2 = System.Runtime.InteropServices.Marshal.GetFunctionPointerForDelegate (d2);

    Console.WriteLine ($"i1 = {i1}, i2 = {i2}");
    GC.KeepAlive (c);
    GC.KeepAlive (d1);
    GC.KeepAlive (d2);
  }
    
}
```

Expected Output:

```
i1 = 12345678, i2 = 24681012
```
(or some other distinct arbitrary pointer values)

Actual Output:
WASM:
```
WASM-ERR: * Assertion at /__w/1/s/src/mono/mono/metadata/marshal.c:399, condition `<disabled>' not met
```

on non-WASM:
```
* Assertion at marshal.c:399, condition `mono_gchandle_target_equal (gchandle, MONO_HANDLE_CAST (MonoObject, d))' not met
```

**Comment by lambdageek** [comment]
One way to test this on non-wasm is to change 

https://github.com/dotnet/runtime/blob/master/src/mono/mono/mini/mini-amd64.h#L468

to 

```c
#undef MONO_ARCH_HAVE_INTERP_NATIVE_TO_MANAGED
```

And then run the above repro with `mono --interp Program.exe`

**Comment by lambdageek** [comment]
Because on platforms without `MONO_ARCH_HAVE_INTERP_NATIVE_TO_MANAGED` the same function is returned for all calls to `interp_create_method_pointer`, any two arbitrary calls to `GetFunctionPointerForDelegate` anywhere in the program will trigger this assertion

**Comment by lateralusX** [comment]
Disabled delegate callbacks in eventpipe library on WASM due to hitting this issue, triggering:

`WASM-ERR: * Assertion at /__w/1/s/src/mono/mono/metadata/marshal.c:399, condition `<disabled>' not met`

When this is fixed on WASM we should re-enable delegate callbacks in eventpipe library src/mono/mono/metadata/icall-eventpipe.c.

**Comment by vargaz** [comment]
This is fixed now by 4647328f7699a11b39297883791fcf7056906e9e.
